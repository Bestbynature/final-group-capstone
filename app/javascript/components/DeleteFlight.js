import React, { useState, useRef, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { deleteFlight, fetchFlights } from '../redux/flights/flightsSlice';
import AddFlight from './AddFlight';

const DeleteFlight = () => {
  const dispatch = useDispatch();
  const flights = useSelector((state) => state.flights.flights);

  const [deletedFlights, setDeletedFlights] = useState([]);
  const deletedFlightRef = useRef(null);

  const handleDelete = (flightId) => {
    const flightToDelete = flights.find((flight) => flight.id === flightId);
    dispatch(deleteFlight(flightId));
    if (flightToDelete) {
      deletedFlightRef.current = flightToDelete;
      setDeletedFlights([...deletedFlights, flightToDelete]);
    }
  };

  useEffect(() => {
    if (flights.length === 0) {
      dispatch(fetchFlights());
    }
  }, [dispatch, flights.length]);

  return (
    <div className="delete-flight">
      <AddFlight />
      {deletedFlights.length > 0 && (
        <ul>
          <h2>Deleted Flights</h2>
          {deletedFlights.map((deletedFlight) => (
            <li key={deletedFlight.id}>
              <span className="deleted">{deletedFlight.name}</span>
              <button className="btn-deleted" type="button" disabled>Delete</button>
            </li>
          ))}
        </ul>
      )}
      {flights.length > 0 ? (
        <ul>
          <h2>Delete Flight</h2>
          {flights.map((flight) => (
            <li key={flight.id} data-testid="flight-item">
              {flight.name}
              <button className="btn-delete" type="button" onClick={() => handleDelete(flight.id)} data-testid="delete-button">Delete</button>
            </li>
          ))}
        </ul>
      ) : (
        <p>No flights available</p>
      )}
    </div>
  );
};

export default DeleteFlight;
