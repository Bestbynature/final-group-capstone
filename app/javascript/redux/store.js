import { configureStore } from '@reduxjs/toolkit';
import flightsReducer from './flights/flightsSlice';

const store = configureStore({
  reducer: {
    flights: flightsReducer,
  },
});

export default store;
