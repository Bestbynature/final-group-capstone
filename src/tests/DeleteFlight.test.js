import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import DeleteFlight from '../../app/javascript/components/DeleteFlight';
import store from '../../app/javascript/redux/store';
import reducer, {
  fetchFlights,
} from '../../app/javascript/redux/flights/flightsSlice';

jest.mock('../../app/javascript/components/AddFlight', () => ({
  __esModule: true,
  default: () => <div data-testid="add-flight-mock">AddFlight Mock</div>,
}));

describe('DeleteFlights', () => {
  it('renders flights correctly', () => {
    const deletePage = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <DeleteFlight />
          </Provider>
        </BrowserRouter>,
      )
      .toJSON();
    expect(deletePage).toMatchSnapshot();
  });

  it('renders Delete page correctly', () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <DeleteFlight />
        </Provider>
      </BrowserRouter>,
    );
  });
});

describe('redux', () => {
  const state = {
    flights: [
      {
        name: 'Egypt', picture: 'https://images.pexels.com/photos/3958516/pexels-photo-3958516.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=200&w=300', base_price: 1100, available_slots: 10, reserved: false,
      },
    ],
    isLoading: false,
    error: '',
  };
  expect(reducer(state, fetchFlights)).toEqual({
    flights: [
      {
        name: 'Egypt', picture: 'https://images.pexels.com/photos/3958516/pexels-photo-3958516.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=200&w=300', base_price: 1100, available_slots: 10, reserved: false,
      },
    ],
    isLoading: false,
    error: '',
  });
});
