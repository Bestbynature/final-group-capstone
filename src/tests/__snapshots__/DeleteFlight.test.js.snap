// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteFlights renders flights correctly 1`] = `
<div
  className="delete-flight"
>
  <div
    data-testid="add-flight-mock"
  >
    AddFlight Mock
  </div>
  <p>
    No flights available
  </p>
</div>
`;
